airbyte-integrations/connectors/destination-redshift/src/main/kotlin/io/airbyte/integrations/destination/redshift/operation/RedshiftStagingStorageOperation.kt

```diff
@@ -0,0 +1,253 @@
+package io.airbyte.integrations.destination.redshift.operation
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import io.airbyte.cdk.integrations.base.JavaBaseConstants
+import io.airbyte.cdk.integrations.destination.record_buffer.SerializableBuffer
+import io.airbyte.cdk.integrations.destination.s3.AesCbcEnvelopeEncryptionBlobDecorator
+import io.airbyte.cdk.integrations.destination.s3.S3DestinationConfig
+import io.airbyte.cdk.integrations.destination.s3.S3StorageOperations
+import io.airbyte.cdk.integrations.destination.staging.SerialStagingConsumerFactory
+import io.airbyte.integrations.base.destination.operation.StorageOperation
+import io.airbyte.integrations.base.destination.typing_deduping.Sql
+import io.airbyte.integrations.base.destination.typing_deduping.StreamConfig
+import io.airbyte.integrations.base.destination.typing_deduping.StreamId
+import io.airbyte.integrations.base.destination.typing_deduping.TyperDeduperUtil
+import io.airbyte.integrations.destination.redshift.RedshiftSQLNameTransformer
+import io.airbyte.integrations.destination.redshift.manifest.Entry
+import io.airbyte.integrations.destination.redshift.manifest.Manifest
+import io.airbyte.integrations.destination.redshift.typing_deduping.RedshiftDestinationHandler
+import io.airbyte.integrations.destination.redshift.typing_deduping.RedshiftSqlGenerator
+import io.airbyte.protocol.models.v0.DestinationSyncMode
+import io.github.oshai.kotlinlogging.KotlinLogging
+import java.time.Instant
+import java.time.ZoneOffset
+import java.time.ZonedDateTime
+import java.util.Base64
+import java.util.Optional
+import java.util.UUID
+import java.util.stream.Collectors
+
+private val log = KotlinLogging.logger {}
+
+/**
+ * @param keyEncryptingKey The KEK to use when writing to S3, or null if encryption is disabled.
+ * If this parameter is nonnull, then `s3StorageOperations` MUST have an
+ * [AesCbcEnvelopeEncryptionBlobDecorator] added (via `s3StorageOperations#addBlobDecorator`).
+ */
+class RedshiftStagingStorageOperation(
+    private val s3Config: S3DestinationConfig,
+    private val keepStagingFiles: Boolean,
+    private val s3StorageOperations: S3StorageOperations,
+    private val sqlGenerator: RedshiftSqlGenerator,
+    private val destinationHandler: RedshiftDestinationHandler,
+    private val keyEncryptingKey: ByteArray?,
+    private val connectionId: UUID = SerialStagingConsumerFactory.RANDOM_CONNECTION_ID,
```

Comment:
Nuke the references of old `SerialStagingConsumerFactory`, easier to delete them later.  use an internal val, unless you choose to inject it for unit tests or something.